app/javascript/components/visualization/ScatterPlot.js

```diff
@@ -45,24 +45,55 @@ function RawScatterPlot({
   const { ErrorComponent, setShowError, setErrorContent } = useErrorMessage()
   /** Process scatter plot data fetched from server */
   function handleResponse(clusterResponse) {
-    const apiOk = checkScpApiResponse(clusterResponse,
+    const [scatter, perfTime] = clusterResponse
+
+    const apiOk = checkScpApiResponse(scatter,
       () => Plotly.purge(graphElementId),
       setShowError,
       setErrorContent)
+
     if (apiOk) {
       // Get Plotly layout
-      const layout = getPlotlyLayout(clusterResponse)
+      const layout = getPlotlyLayout(scatter)
       const { width, height } = dimensions
       layout.width = width
       layout.height = height
-      formatMarkerColors(clusterResponse.data, clusterResponse.annotParams.type, clusterResponse.gene)
-      formatHoverLabels(clusterResponse.data, clusterResponse.annotParams.type, clusterResponse.gene)
-      const dataScatterColor = processTraceScatterColor(clusterResponse.data, scatterColor)
-      Plotly.newPlot(graphElementId, clusterResponse.data, layout)
+      formatMarkerColors(scatter.data, scatter.annotParams.type, scatter.gene)
+      formatHoverLabels(scatter.data, scatter.annotParams.type, scatter.gene)
+      const dataScatterColor = processTraceScatterColor(scatter.data, scatterColor)
+
+      const start = `perfTimeFrontendStart-${graphElementId}`
+      performance.mark(start)
+
+      Plotly.newPlot(graphElementId, scatter.data, layout)
+
+      const end = `perfTimeFrontendEnd-${graphElementId}`
+      performance.mark(end)
+      const perfTimeFrontend = performance.measure(graphElementId, start, end).duration
```

Comment:
I think moving these performance-related lines after newPlot into a helper logPerformance() method would improve the readability of this handleResponse.

Reply:
Agreed, done in a517ef9.